{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import pandas as pd \n",
    "\n",
    "class TF_Data_Processing():\n",
    "    \n",
    "    def __init__(self,dir_path,cancer,select_slice_path,save_path):\n",
    "\n",
    "        self.cancer=cancer\n",
    "        self.dir_path=dir_path\n",
    "        self.save_path=save_path\n",
    "        self.select_slice_path=select_slice_path\n",
    "        self.file_name=os.listdir(self.dir_path)\n",
    "        self.file_path= [os.path.join(self.dir_path,i) for i in os.listdir(self.dir_path)]\n",
    "    \n",
    "\n",
    "\n",
    "    def Create_folder(self):\n",
    "\n",
    "        if not os.path.exists(self.save_path):\n",
    "            os.makedirs(self.save_path)\n",
    "\n",
    "\n",
    "\n",
    "    def Single_slice_data_extraction(self):\n",
    "\n",
    "        all_slice=pd.read_csv(self.select_slice_path)[\"slice\"].values\n",
    "        for i in self.file_name:\n",
    "            cancer=re.findall(r'\\D+', i.split(\"-\")[0])\n",
    "            slice=i.split(\".loom.tsv\")[0]\n",
    "            if cancer[0] == self.cancer and slice in all_slice:\n",
    "                data=pd.read_csv(os.path.join(self.dir_path,i),sep=\"\\t\")\n",
    "                data[\"TF\"].value_counts().to_csv(self.save_path+\"\\\\\"+slice+\".csv\",columns=[\"TF\"])\n",
    "    \n",
    "\n",
    "\n",
    "    def TF_slice_num_statistics(self):\n",
    "\n",
    "        TF_file_path= [os.path.join(self.save_path,i) for i in os.listdir(self.save_path)]\n",
    "        dfs = []\n",
    "        for file in TF_file_path:\n",
    "            cancer = re.findall(self.cancer, file)\n",
    "            if cancer[0]==self.cancer:\n",
    "                df = pd.read_table(file,sep=\",\",index_col=0)\n",
    "                dfs.append(df)\n",
    "        combined_df = pd.concat(dfs,axis=0).index.value_counts()\n",
    "        combined_df=pd.DataFrame(combined_df)\n",
    "        combined_df.columns=[\"slice_num\"]\n",
    "        combined_df[\"all_slice_num\"]=[len(TF_file_path) for i in range(len(combined_df.index))]\n",
    "        combined_df[\"ratio\"]=(combined_df[\"slice_num\"]/combined_df[\"all_slice_num\"]).round(3)\n",
    "        combined_df.to_csv(os.path.join(self.save_path,\"TF-slice-num.txt\"),sep=\"\\t\")\n",
    "        print(f\"{self.cancer}的TF-slice数目统计完成了!\")\n",
    "        print(\"******************************************************************************************************************\")\n",
    "        print(combined_df)\n",
    "\n",
    "\n",
    "\n",
    "    def Drop_single_slice_file(self):\n",
    "\n",
    "        files=os.listdir(self.save_path)\n",
    "        for file in files:\n",
    "            cancer=re.findall(r'\\D+', file.split(\"-\")[0])\n",
    "            if cancer[0]==self.cancer:\n",
    "                file_path = os.path.join(self.save_path, file)\n",
    "                os.remove(file_path)\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "    def TF_terget_statistics(self):\n",
    "\n",
    "        dfs = []\n",
    "        select_slice=pd.read_csv(self.select_slice_path)[\"slice\"].values\n",
    "        for i in self.file_name:\n",
    "            cancer=re.findall(r'\\D+', i.split(\"-\")[0])\n",
    "            slice=i.split(\".loom.tsv\")[0]\n",
    "            if cancer[0] == self.cancer and slice in select_slice:\n",
    "                data=pd.read_csv(os.path.join(self.dir_path,i),sep=\"\\t\")\n",
    "                data[\"slice\"]=[slice for i in range(len(data.index))]\n",
    "                data[\"cancer\"]=[cancer[0] for i in range(len(data.index))]\n",
    "                dfs.append(data)\n",
    "        combined_df = pd.concat(dfs)\n",
    "        group_list=combined_df.groupby([\"TF\",\"slice\"])\n",
    "        num=0\n",
    "        table=pd.DataFrame(columns=[\"TF\",\"target\",\"target_num\",\"importance\",\"slice\",\"cancer\"])\n",
    "        for group_name, group_data in group_list:\n",
    "            target=[i for i in group_data[\"target\"]]\n",
    "            importance=[str(i) for i in group_data[\"importance\"].round(3)]\n",
    "    \n",
    "            cancer=[i for i in group_data[\"cancer\"]]\n",
    "            target_num=len(target)\n",
    "            table.loc[num,\"TF\"]=group_name[0]\n",
    "            table.loc[num,\"target\"]=\",\".join(target)\n",
    "            table.loc[num,\"target_num\"]=target_num\n",
    "            table.loc[num,\"importance\"]=\",\".join(importance)\n",
    "            table.loc[num,\"slice\"]=group_name[1].replace(\"-\",\"_\")\n",
    "            table.loc[num,\"cancer\"]=self.cancer\n",
    "            num+=1\n",
    "        table.to_csv(self.save_path+\"\\\\TF-Statistics\"+\".txt\",sep=\"\\t\",index=False)\n",
    "        print(f\"{self.cancer}的TF-target统计完成了!\")\n",
    "        print(\"******************************************************************************************************************\")\n",
    "        print(table)\n",
    "        \n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "开始hn_as的TF的数据统计\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-238-c4190c38172b>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      5\u001b[0m     \u001b[0mtest\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mCreate_folder\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      6\u001b[0m     \u001b[0mtest\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mSingle_slice_data_extraction\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 7\u001b[1;33m     \u001b[0mtest\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mTF_slice_num_statistics\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      8\u001b[0m     \u001b[0mtest\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mDrop_single_slice_file\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      9\u001b[0m     \u001b[0mtest\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mTF_terget_statistics\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-237-e39d18c26653>\u001b[0m in \u001b[0;36mTF_slice_num_statistics\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m     41\u001b[0m         \u001b[1;32mfor\u001b[0m \u001b[0mfile\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mTF_file_path\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     42\u001b[0m             \u001b[0mcancer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mre\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfindall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcancer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mfile\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 43\u001b[1;33m             \u001b[1;32mif\u001b[0m \u001b[0mcancer\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m==\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcancer\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     44\u001b[0m                 \u001b[0mdf\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_table\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0msep\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\",\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mindex_col\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     45\u001b[0m                 \u001b[0mdfs\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "cancer=[\"hn_as\"]\n",
    "for i in cancer:\n",
    "    print(f\"开始{i}的TF的数据统计\")\n",
    "    test=TF_Data_Processing(dir_path=\"result_grn\",cancer=i,select_slice_path=\"E:\\苏师兄代码和文件\\select_slice1.csv\",save_path=i.upper())\n",
    "    test.Create_folder()\n",
    "    test.Single_slice_data_extraction()\n",
    "    test.TF_slice_num_statistics()\n",
    "    test.Drop_single_slice_file()\n",
    "    test.TF_terget_statistics()\n",
    "print(\"所有统计完成\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#总表统计\n",
    "import os\n",
    "import pandas as pd\n",
    "dfs=[]\n",
    "for i in os.listdir(\"E:\\苏师兄代码和文件\\pyscenic\"):\n",
    "    if os.path.exists(os.path.join(r\"E:\\苏师兄代码和文件\\pyscenic\",i+\"\\\\TF-slice-num.txt\")):\n",
    "        data=pd.read_csv(os.path.join(r\"E:\\苏师兄代码和文件\\pyscenic\",i+\"\\\\TF-slice-num.txt\"),sep=\"\\t\")\n",
    "        dfs.append(data)\n",
    "combined_df = pd.concat(dfs)\n",
    "combined_df = combined_df.rename(columns={combined_df.columns[0]: \"TF\"})\n",
    "table=pd.DataFrame()\n",
    "table[\"TF\"]=combined_df.groupby(\"TF\")[\"slice_num\"].sum().index\n",
    "table[\"slice_num\"]=combined_df.groupby(\"TF\")[\"slice_num\"].sum().values\n",
    "table[\"all_slice_num\"]=[144 for i in range(len(table[\"TF\"]))]\n",
    "table[\"slice_ratio\"]=(table[\"slice_num\"]/table[\"all_slice_num\"]).round(3)\n",
    "table[\"cancer_num\"]=combined_df[\"TF\"].value_counts().values\n",
    "table[\"all_cancer_num\"]=[18 for i in range(len(table[\"TF\"]))]\n",
    "table[\"cancer_ratio\"]=(table[\"cancer_num\"]/table[\"all_cancer_num\"]).round(3)\n",
    "table.to_csv(\"ALL\\\\All-TF-slice-num.txt\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>slice_num</th>\n",
       "      <th>all_slice_num</th>\n",
       "      <th>slice_ratio</th>\n",
       "      <th>cancer_num</th>\n",
       "      <th>all_cancer_num</th>\n",
       "      <th>cancer_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>45</td>\n",
       "      <td>144</td>\n",
       "      <td>0.312</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCF2</td>\n",
       "      <td>114</td>\n",
       "      <td>144</td>\n",
       "      <td>0.792</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>143</td>\n",
       "      <td>144</td>\n",
       "      <td>0.993</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACAA1</td>\n",
       "      <td>143</td>\n",
       "      <td>144</td>\n",
       "      <td>0.993</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACO1</td>\n",
       "      <td>143</td>\n",
       "      <td>144</td>\n",
       "      <td>0.993</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1815</th>\n",
       "      <td>ZSWIM1</td>\n",
       "      <td>139</td>\n",
       "      <td>144</td>\n",
       "      <td>0.965</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>0.111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1816</th>\n",
       "      <td>ZXDA</td>\n",
       "      <td>106</td>\n",
       "      <td>144</td>\n",
       "      <td>0.736</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>0.111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1817</th>\n",
       "      <td>ZXDB</td>\n",
       "      <td>138</td>\n",
       "      <td>144</td>\n",
       "      <td>0.958</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>0.111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1818</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>143</td>\n",
       "      <td>144</td>\n",
       "      <td>0.993</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>0.056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1819</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>2</td>\n",
       "      <td>144</td>\n",
       "      <td>0.014</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>0.056</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1820 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          TF  slice_num  all_slice_num  slice_ratio  cancer_num  \\\n",
       "0       A1CF         45            144        0.312          18   \n",
       "1      ABCF2        114            144        0.792          18   \n",
       "2       ABL1        143            144        0.993          18   \n",
       "3      ACAA1        143            144        0.993          18   \n",
       "4       ACO1        143            144        0.993          18   \n",
       "...      ...        ...            ...          ...         ...   \n",
       "1815  ZSWIM1        139            144        0.965           2   \n",
       "1816    ZXDA        106            144        0.736           2   \n",
       "1817    ZXDB        138            144        0.958           2   \n",
       "1818    ZXDC        143            144        0.993           1   \n",
       "1819    ZZZ3          2            144        0.014           1   \n",
       "\n",
       "      all_cancer_num  cancer_ratio  \n",
       "0                 18         1.000  \n",
       "1                 18         1.000  \n",
       "2                 18         1.000  \n",
       "3                 18         1.000  \n",
       "4                 18         1.000  \n",
       "...              ...           ...  \n",
       "1815              18         0.111  \n",
       "1816              18         0.111  \n",
       "1817              18         0.111  \n",
       "1818              18         0.056  \n",
       "1819              18         0.056  \n",
       "\n",
       "[1820 rows x 7 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#总表统计\n",
    "import os\n",
    "import pandas as pd\n",
    "dfs=[]\n",
    "for i in os.listdir(\"E:\\苏师兄代码和文件\\pyscenic\"):\n",
    "    if os.path.exists(os.path.join(r\"E:\\苏师兄代码和文件\\pyscenic\",i+\"\\\\TF-Statistics.txt\")):\n",
    "        data=pd.read_csv(os.path.join(r\"E:\\苏师兄代码和文件\\pyscenic\",i+\"\\\\TF-Statistics.txt\"),sep=\"\\t\")\n",
    "        dfs.append(data)\n",
    "combined_df = pd.concat(dfs)\n",
    "combined_df = combined_df.rename(columns={combined_df.columns[0]: \"TF\"})\n",
    "combined_df=combined_df.sort_values(by='TF', ascending=True)\n",
    "combined_df.to_csv(\"ALL\\\\All-TF-Statistics.txt\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>target</th>\n",
       "      <th>target_num</th>\n",
       "      <th>importance</th>\n",
       "      <th>slice</th>\n",
       "      <th>cancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>DRC3,YOD1,AC147651.1,CHRNA4,TAB1,FAM242C,ANKFY...</td>\n",
       "      <td>370</td>\n",
       "      <td>1.8,1.472,1.384,1.374,1.093,1.09,1.066,0.945,0...</td>\n",
       "      <td>crc10_slice3</td>\n",
       "      <td>crc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>SLC4A9,LINC02577,CHAC1,AL365181.4,AC004241.3,Z...</td>\n",
       "      <td>3035</td>\n",
       "      <td>5.043,4.324,3.806,3.623,3.399,2.943,2.917,2.87...</td>\n",
       "      <td>lihc03_slice2</td>\n",
       "      <td>lihc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>LINC01339,AC004593.2,ZNF607,LINC02518,AL355297...</td>\n",
       "      <td>2241</td>\n",
       "      <td>5.412,4.368,3.977,3.608,3.297,2.84,2.804,2.482...</td>\n",
       "      <td>lihc03_slice1</td>\n",
       "      <td>lihc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>VPREB3,C1orf52,SNCB,AL354919.1,ZNF417,LTBP1,US...</td>\n",
       "      <td>2671</td>\n",
       "      <td>2.59,2.509,2.397,2.165,2.103,1.984,1.969,1.959...</td>\n",
       "      <td>lihc02_slice9</td>\n",
       "      <td>lihc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>TMPRSS15,PLA2G6,MPP7,ZBTB47,SETBP1,TNXB,TRBJ1-...</td>\n",
       "      <td>223</td>\n",
       "      <td>1.174,1.103,0.972,0.811,0.698,0.557,0.552,0.46...</td>\n",
       "      <td>pdac03_slice2</td>\n",
       "      <td>pdac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52172</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>RAB33B,DUSP9,IPO8,POLH,HSF4,SELPLG,CCNC,SNHG18...</td>\n",
       "      <td>360</td>\n",
       "      <td>2.204,2.061,1.939,1.894,1.88,1.815,1.647,1.646...</td>\n",
       "      <td>gbm05_slice14</td>\n",
       "      <td>gbm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52163</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>DNAAF3,AC010931.3,CPLX3,LCE1D,PARD6G,ZNF264,AL...</td>\n",
       "      <td>1004</td>\n",
       "      <td>3.096,2.187,1.728,1.382,1.245,1.112,1.104,1.03...</td>\n",
       "      <td>gbm04_slice4</td>\n",
       "      <td>gbm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2849</th>\n",
       "      <td>ZXDC</td>\n",
       "      <td>CRISP1,TULP2,PLA2G2C,LDOC1,SPSB4,KRTAP26-1,ZNF...</td>\n",
       "      <td>783</td>\n",
       "      <td>2.611,1.238,1.004,1.003,0.927,0.914,0.86,0.846...</td>\n",
       "      <td>skcm13_slice1</td>\n",
       "      <td>skcm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2479</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>SLC6A1,MAFTRR,PDS5B,LOC103344931,ST6GALNAC3,AR...</td>\n",
       "      <td>748</td>\n",
       "      <td>5.038,3.731,3.717,3.077,2.848,2.224,2.072,1.74...</td>\n",
       "      <td>rcc01_slice2</td>\n",
       "      <td>rcc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2478</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>STAP2,ZKSCAN2,HCG27,TGFBRAP1,NR2C2,RAD52,SUPT5...</td>\n",
       "      <td>885</td>\n",
       "      <td>4.031,3.4,3.061,2.704,2.496,2.131,2.098,2.014,...</td>\n",
       "      <td>rcc01_slice1</td>\n",
       "      <td>rcc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204701 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         TF                                             target  target_num  \\\n",
       "10     A1CF  DRC3,YOD1,AC147651.1,CHRNA4,TAB1,FAM242C,ANKFY...         370   \n",
       "14     A1CF  SLC4A9,LINC02577,CHAC1,AL365181.4,AC004241.3,Z...        3035   \n",
       "13     A1CF  LINC01339,AC004593.2,ZNF607,LINC02518,AL355297...        2241   \n",
       "12     A1CF  VPREB3,C1orf52,SNCB,AL354919.1,ZNF417,LTBP1,US...        2671   \n",
       "1      A1CF  TMPRSS15,PLA2G6,MPP7,ZBTB47,SETBP1,TNXB,TRBJ1-...         223   \n",
       "...     ...                                                ...         ...   \n",
       "52172  ZXDC  RAB33B,DUSP9,IPO8,POLH,HSF4,SELPLG,CCNC,SNHG18...         360   \n",
       "52163  ZXDC  DNAAF3,AC010931.3,CPLX3,LCE1D,PARD6G,ZNF264,AL...        1004   \n",
       "2849   ZXDC  CRISP1,TULP2,PLA2G2C,LDOC1,SPSB4,KRTAP26-1,ZNF...         783   \n",
       "2479   ZZZ3  SLC6A1,MAFTRR,PDS5B,LOC103344931,ST6GALNAC3,AR...         748   \n",
       "2478   ZZZ3  STAP2,ZKSCAN2,HCG27,TGFBRAP1,NR2C2,RAD52,SUPT5...         885   \n",
       "\n",
       "                                              importance          slice cancer  \n",
       "10     1.8,1.472,1.384,1.374,1.093,1.09,1.066,0.945,0...   crc10_slice3    crc  \n",
       "14     5.043,4.324,3.806,3.623,3.399,2.943,2.917,2.87...  lihc03_slice2   lihc  \n",
       "13     5.412,4.368,3.977,3.608,3.297,2.84,2.804,2.482...  lihc03_slice1   lihc  \n",
       "12     2.59,2.509,2.397,2.165,2.103,1.984,1.969,1.959...  lihc02_slice9   lihc  \n",
       "1      1.174,1.103,0.972,0.811,0.698,0.557,0.552,0.46...  pdac03_slice2   pdac  \n",
       "...                                                  ...            ...    ...  \n",
       "52172  2.204,2.061,1.939,1.894,1.88,1.815,1.647,1.646...  gbm05_slice14    gbm  \n",
       "52163  3.096,2.187,1.728,1.382,1.245,1.112,1.104,1.03...   gbm04_slice4    gbm  \n",
       "2849   2.611,1.238,1.004,1.003,0.927,0.914,0.86,0.846...  skcm13_slice1   skcm  \n",
       "2479   5.038,3.731,3.717,3.077,2.848,2.224,2.072,1.74...   rcc01_slice2    rcc  \n",
       "2478   4.031,3.4,3.061,2.704,2.496,2.131,2.098,2.014,...   rcc01_slice1    rcc  \n",
       "\n",
       "[204701 rows x 6 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[\"cancer\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MBD1       143\n",
       "CBFA2T2    143\n",
       "ZNF316     143\n",
       "DNMT3A     143\n",
       "ZNF397     143\n",
       "          ... \n",
       "ZNF735       3\n",
       "FOXD4L4      2\n",
       "TGIF2LY      2\n",
       "ZZZ3         2\n",
       "SSX2         1\n",
       "Name: TF, Length: 1820, dtype: int64"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[\"TF\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import  ast\n",
    "import pandas as pd "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data=pd.read_csv(r\"E:\\苏师兄代码和文件\\pyscenic\\result_ctx\\reg_brca01-slice1.loom.tsv.csv\")\n",
    "data.columns=data.iloc[0,:]\n",
    "data=data[1:]\n",
    "data.columns.values[0]=\"TF\"\n",
    "data.columns.values[1]=\"MotifID\"\n",
    "data=data[1:]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Target_num=[]\n",
    "for i in data[\"TargetGenes\"]:\n",
    "    pattern = re.compile('[a-zA-Z]+')\n",
    "    english_words = [re.findall(pattern, item[0])[0] for item in ast.literal_eval(i)]\n",
    "    Target_num.append(len(english_words))\n",
    "data[\"Target_num\"]=Target_num\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>MotifID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>NES</th>\n",
       "      <th>MotifSimilarityQvalue</th>\n",
       "      <th>OrthologousIdentity</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Context</th>\n",
       "      <th>TargetGenes</th>\n",
       "      <th>RankAtMax</th>\n",
       "      <th>Target_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>XBP1</td>\n",
       "      <td>metacluster_71.16</td>\n",
       "      <td>0.04442817365320294</td>\n",
       "      <td>3.186038904705333</td>\n",
       "      <td>9.44332e-08</td>\n",
       "      <td>1.0</td>\n",
       "      <td>motif similar to taipale_cyt_meth__XBP1_NKMCAC...</td>\n",
       "      <td>frozenset({'activating', 'weight&gt;75.0%', 'hg38...</td>\n",
       "      <td>[('CNOT2', 0.4836674489060573), ('PEX2', 1.661...</td>\n",
       "      <td>4970</td>\n",
       "      <td>529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>675</th>\n",
       "      <td>ELF3</td>\n",
       "      <td>metacluster_191.2</td>\n",
       "      <td>0.04968564357473354</td>\n",
       "      <td>3.1943287350642726</td>\n",
       "      <td>6.90333e-06</td>\n",
       "      <td>1.0</td>\n",
       "      <td>motif similar to taipale_cyt_meth__ELF3_ACCCGG...</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('SOCS6', 9.443884364310829e-08), ('SERTAD2',...</td>\n",
       "      <td>4905</td>\n",
       "      <td>316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742</th>\n",
       "      <td>STAT1</td>\n",
       "      <td>tfdimers__MD00523</td>\n",
       "      <td>0.0435436743791873</td>\n",
       "      <td>3.182879452680371</td>\n",
       "      <td>1.91863e-09</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is annotated for similar motif tfdimers__...</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('FBLN7', 0.4067316237631928), ('SHC2', 0.100...</td>\n",
       "      <td>4955</td>\n",
       "      <td>316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>738</th>\n",
       "      <td>SPDEF</td>\n",
       "      <td>taipale_tf_pairs__GCM1_ELK1_RTGCGGGCGGAAGTN_CAP_1</td>\n",
       "      <td>0.044129575396384936</td>\n",
       "      <td>3.0311327308119522</td>\n",
       "      <td>3.95882e-08</td>\n",
       "      <td>1.0</td>\n",
       "      <td>motif similar to taipale_tf_pairs__GCM1_SPDEF_...</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('ACSL4', 0.085369573176543), ('SLC45A4', 0.4...</td>\n",
       "      <td>4737</td>\n",
       "      <td>298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>737</th>\n",
       "      <td>SPDEF</td>\n",
       "      <td>metacluster_166.2</td>\n",
       "      <td>0.04830189860628467</td>\n",
       "      <td>3.793683640426003</td>\n",
       "      <td>1.83208e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>motif similar to taipale_tf_pairs__ELK1_SPDEF_...</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('NPRL3', 1.1261173811912362), ('C1GALT1', 0....</td>\n",
       "      <td>4973</td>\n",
       "      <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>PPARG</td>\n",
       "      <td>tfdimers__MD00172</td>\n",
       "      <td>0.11869592741084617</td>\n",
       "      <td>3.0763678507109002</td>\n",
       "      <td>5.77514e-08</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is annotated for similar motif tfdimers__...</td>\n",
       "      <td>frozenset({'top5perTarget', 'activating', 'hg3...</td>\n",
       "      <td>[('SORBS3', 0.9420932782059972), ('DEDD', 0.59...</td>\n",
       "      <td>88</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>CLOCK</td>\n",
       "      <td>taipale_tf_pairs__CLOCK_BHLHA15_NCACGTGNNNNNCA...</td>\n",
       "      <td>0.155480059084195</td>\n",
       "      <td>4.447664461124356</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'top5perTarget', 'activating', 'hg3...</td>\n",
       "      <td>[('NUP35', 0.7917121613195244), ('CAMK2N1', 1....</td>\n",
       "      <td>53</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>HOXC4</td>\n",
       "      <td>metacluster_9.9</td>\n",
       "      <td>0.10955194485475135</td>\n",
       "      <td>3.198181842922439</td>\n",
       "      <td>6.45702e-06</td>\n",
       "      <td>1.0</td>\n",
       "      <td>motif similar to swissregulon__hs__HOXC4 ('hs_...</td>\n",
       "      <td>frozenset({'activating', 'hg38_10kbp_up_10kbp_...</td>\n",
       "      <td>[('KDSR', 0.9848095898039196), ('HOXC4', 1.0)]</td>\n",
       "      <td>71</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ATF3</td>\n",
       "      <td>tfdimers__MD00565</td>\n",
       "      <td>0.14307401936648614</td>\n",
       "      <td>3.4502281215356465</td>\n",
       "      <td>1.54335e-06</td>\n",
       "      <td>1.0</td>\n",
       "      <td>motif similar to tfdimers__MD00119 ('M00973_re...</td>\n",
       "      <td>frozenset({'top5perTarget', 'activating', 'hg3...</td>\n",
       "      <td>[('EFNA3', 1.0427155236199108), ('ZNF292', 0.7...</td>\n",
       "      <td>51</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>TBX3</td>\n",
       "      <td>taipale_tf_pairs__MGA_DLX2_AGGTGNTAATTR_CAP</td>\n",
       "      <td>0.11929565420197466</td>\n",
       "      <td>3.131820836693068</td>\n",
       "      <td>6.64197e-06</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is annotated for similar motif taipale_tf...</td>\n",
       "      <td>frozenset({'top5perTarget', 'activating', 'hg3...</td>\n",
       "      <td>[('TBX3', 1.0), ('MAP1A', 1.0395421613059834)]</td>\n",
       "      <td>164</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>758 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0       TF                                            MotifID  \\\n",
       "101   XBP1                                  metacluster_71.16   \n",
       "675   ELF3                                  metacluster_191.2   \n",
       "742  STAT1                                  tfdimers__MD00523   \n",
       "738  SPDEF  taipale_tf_pairs__GCM1_ELK1_RTGCGGGCGGAAGTN_CAP_1   \n",
       "737  SPDEF                                  metacluster_166.2   \n",
       "..     ...                                                ...   \n",
       "498  PPARG                                  tfdimers__MD00172   \n",
       "416  CLOCK  taipale_tf_pairs__CLOCK_BHLHA15_NCACGTGNNNNNCA...   \n",
       "186  HOXC4                                    metacluster_9.9   \n",
       "406   ATF3                                  tfdimers__MD00565   \n",
       "518   TBX3        taipale_tf_pairs__MGA_DLX2_AGGTGNTAATTR_CAP   \n",
       "\n",
       "0                     AUC                 NES MotifSimilarityQvalue  \\\n",
       "101   0.04442817365320294   3.186038904705333           9.44332e-08   \n",
       "675   0.04968564357473354  3.1943287350642726           6.90333e-06   \n",
       "742    0.0435436743791873   3.182879452680371           1.91863e-09   \n",
       "738  0.044129575396384936  3.0311327308119522           3.95882e-08   \n",
       "737   0.04830189860628467   3.793683640426003           1.83208e-07   \n",
       "..                    ...                 ...                   ...   \n",
       "498   0.11869592741084617  3.0763678507109002           5.77514e-08   \n",
       "416     0.155480059084195   4.447664461124356                   0.0   \n",
       "186   0.10955194485475135   3.198181842922439           6.45702e-06   \n",
       "406   0.14307401936648614  3.4502281215356465           1.54335e-06   \n",
       "518   0.11929565420197466   3.131820836693068           6.64197e-06   \n",
       "\n",
       "0   OrthologousIdentity                                         Annotation  \\\n",
       "101                 1.0  motif similar to taipale_cyt_meth__XBP1_NKMCAC...   \n",
       "675                 1.0  motif similar to taipale_cyt_meth__ELF3_ACCCGG...   \n",
       "742                 1.0  gene is annotated for similar motif tfdimers__...   \n",
       "738                 1.0  motif similar to taipale_tf_pairs__GCM1_SPDEF_...   \n",
       "737                 1.0  motif similar to taipale_tf_pairs__ELK1_SPDEF_...   \n",
       "..                  ...                                                ...   \n",
       "498                 1.0  gene is annotated for similar motif tfdimers__...   \n",
       "416                 1.0                         gene is directly annotated   \n",
       "186                 1.0  motif similar to swissregulon__hs__HOXC4 ('hs_...   \n",
       "406                 1.0  motif similar to tfdimers__MD00119 ('M00973_re...   \n",
       "518                 1.0  gene is annotated for similar motif taipale_tf...   \n",
       "\n",
       "0                                              Context  \\\n",
       "101  frozenset({'activating', 'weight>75.0%', 'hg38...   \n",
       "675  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "742  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "738  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "737  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "..                                                 ...   \n",
       "498  frozenset({'top5perTarget', 'activating', 'hg3...   \n",
       "416  frozenset({'top5perTarget', 'activating', 'hg3...   \n",
       "186  frozenset({'activating', 'hg38_10kbp_up_10kbp_...   \n",
       "406  frozenset({'top5perTarget', 'activating', 'hg3...   \n",
       "518  frozenset({'top5perTarget', 'activating', 'hg3...   \n",
       "\n",
       "0                                          TargetGenes RankAtMax  Target_num  \n",
       "101  [('CNOT2', 0.4836674489060573), ('PEX2', 1.661...      4970         529  \n",
       "675  [('SOCS6', 9.443884364310829e-08), ('SERTAD2',...      4905         316  \n",
       "742  [('FBLN7', 0.4067316237631928), ('SHC2', 0.100...      4955         316  \n",
       "738  [('ACSL4', 0.085369573176543), ('SLC45A4', 0.4...      4737         298  \n",
       "737  [('NPRL3', 1.1261173811912362), ('C1GALT1', 0....      4973         294  \n",
       "..                                                 ...       ...         ...  \n",
       "498  [('SORBS3', 0.9420932782059972), ('DEDD', 0.59...        88           2  \n",
       "416  [('NUP35', 0.7917121613195244), ('CAMK2N1', 1....        53           2  \n",
       "186     [('KDSR', 0.9848095898039196), ('HOXC4', 1.0)]        71           2  \n",
       "406  [('EFNA3', 1.0427155236199108), ('ZNF292', 0.7...        51           2  \n",
       "518     [('TBX3', 1.0), ('MAP1A', 1.0395421613059834)]       164           2  \n",
       "\n",
       "[758 rows x 11 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data=data.sort_values(by=\"Target_num\",ascending=False)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancerlist=[\"brca\",\"cesc\",\"crc\",\"cscc\",\"gbm\",\"gist\",\"hn_as\",\"ipnm\",\"lihc\",\"luad\",\"mibc\",\"oscc\",\"ovca\",\"pcnsl\",\"pdac\",\"prad\",\"rcc\",\"skcm\"][0]\n",
    "dir_path=\"result_ctx\"\n",
    "select_slice_path=\"E:\\苏师兄代码和文件\\select_slice1.csv\"\n",
    "file_name=os.listdir(dir_path)\n",
    "file_path= [os.path.join(dir_path,i) for i in os.listdir(dir_path)]\n",
    "dfs = []\n",
    "select_slice=pd.read_csv(r\"E:\\苏师兄代码和文件\\select_slice1.csv\")[\"slice\"].values\n",
    "for i in file_name:\n",
    "    cancer=re.findall(r'\\D+', i.split(\"-\")[0])[0].replace(\"reg_\",\"\")\n",
    "    slice=i.split(\".loom.tsv.csv\")[0].replace(\"reg_\",\"\")\n",
    "    if cancer == \"brca\" and slice in select_slice:\n",
    "        data=pd.read_csv(r\"E:\\\\苏师兄代码和文件\\\\pyscenic\\\\result_ctx\\\\\"+i)\n",
    "        data.columns=data.iloc[0,:]\n",
    "        data=data[1:]\n",
    "        data.columns.values[0]=\"TF\"\n",
    "        data.columns.values[1]=\"MotifID\"\n",
    "        data=data[1:]\n",
    "        data[\"slice\"]=[slice for i in range(len(data.index))]\n",
    "        data[\"cancer\"]=[cancer for i in range(len(data.index))]\n",
    "        dfs.append(data)\n",
    "combined_df = pd.concat(dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>MotifID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>NES</th>\n",
       "      <th>MotifSimilarityQvalue</th>\n",
       "      <th>OrthologousIdentity</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Context</th>\n",
       "      <th>TargetGenes</th>\n",
       "      <th>RankAtMax</th>\n",
       "      <th>slice</th>\n",
       "      <th>cancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATF4</td>\n",
       "      <td>metacluster_148.1</td>\n",
       "      <td>0.04535987163002242</td>\n",
       "      <td>3.094187231280473</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...</td>\n",
       "      <td>[('CYHR1', 0.7961399861048666), ('ATF3', 0.549...</td>\n",
       "      <td>4316</td>\n",
       "      <td>brca01-slice3</td>\n",
       "      <td>brca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ATF7</td>\n",
       "      <td>tfdimers__MD00403</td>\n",
       "      <td>0.07734985769355478</td>\n",
       "      <td>3.3327772737947217</td>\n",
       "      <td>2.25962e-06</td>\n",
       "      <td>1.0</td>\n",
       "      <td>motif similar to tfdimers__MD00119 ('M00973_re...</td>\n",
       "      <td>frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...</td>\n",
       "      <td>[('TBL1XR1', 0.3794181419999426), ('TP73', 0.9...</td>\n",
       "      <td>2860</td>\n",
       "      <td>brca01-slice3</td>\n",
       "      <td>brca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRF1</td>\n",
       "      <td>metacluster_172.30</td>\n",
       "      <td>0.07246964091972653</td>\n",
       "      <td>3.393020859843327</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...</td>\n",
       "      <td>[('APEH', 0.4085328157090424), ('KIAA0930', 0....</td>\n",
       "      <td>3481</td>\n",
       "      <td>brca01-slice3</td>\n",
       "      <td>brca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CEBPD</td>\n",
       "      <td>metacluster_188.6</td>\n",
       "      <td>0.08619285687232511</td>\n",
       "      <td>3.0759202055964154</td>\n",
       "      <td>4.11114e-06</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is annotated for similar motif tfdimers__...</td>\n",
       "      <td>frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...</td>\n",
       "      <td>[('PPM1H', 2.3367527585852383), ('WRAP73', 0.4...</td>\n",
       "      <td>920</td>\n",
       "      <td>brca01-slice3</td>\n",
       "      <td>brca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CEBPD</td>\n",
       "      <td>metacluster_127.2</td>\n",
       "      <td>0.08783092830360187</td>\n",
       "      <td>3.174015480466356</td>\n",
       "      <td>4.30447e-08</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is annotated for similar motif tfdimers__...</td>\n",
       "      <td>frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...</td>\n",
       "      <td>[('WRAP73', 0.4242448518158827), ('HERPUD2', 0...</td>\n",
       "      <td>573</td>\n",
       "      <td>brca01-slice3</td>\n",
       "      <td>brca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1247</th>\n",
       "      <td>ZNF704</td>\n",
       "      <td>metacluster_144.3</td>\n",
       "      <td>0.052477081428336965</td>\n",
       "      <td>3.451496379493853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('ZNF704', 1.0), ('SIAH1', 0.2095099932575139...</td>\n",
       "      <td>4645</td>\n",
       "      <td>brca26-slice1</td>\n",
       "      <td>brca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1248</th>\n",
       "      <td>ZNF707</td>\n",
       "      <td>jaspar__MA1715.1</td>\n",
       "      <td>0.06133502145318984</td>\n",
       "      <td>3.2644445065952743</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('PIK3R1', 0.2066647993658202), ('ITSN2', 0.1...</td>\n",
       "      <td>3020</td>\n",
       "      <td>brca26-slice1</td>\n",
       "      <td>brca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1249</th>\n",
       "      <td>ZNF846</td>\n",
       "      <td>metacluster_133.2</td>\n",
       "      <td>0.06818181818181818</td>\n",
       "      <td>3.193544830125133</td>\n",
       "      <td>2.55708e-06</td>\n",
       "      <td>1.0</td>\n",
       "      <td>motif similar to kznf__ZNF846_Imbeault2017_RP_...</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('CLSTN1', 0.3020024914065943), ('CCDC174', 0...</td>\n",
       "      <td>356</td>\n",
       "      <td>brca26-slice1</td>\n",
       "      <td>brca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1250</th>\n",
       "      <td>ZNF846</td>\n",
       "      <td>kznf__ZNF846_Imbeault2017_OM_MEME</td>\n",
       "      <td>0.0662720260209182</td>\n",
       "      <td>3.025576739632648</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('SECTM1', 0.4894350004497896), ('CDKL5', 0.5...</td>\n",
       "      <td>381</td>\n",
       "      <td>brca26-slice1</td>\n",
       "      <td>brca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1251</th>\n",
       "      <td>ZNF846</td>\n",
       "      <td>kznf__ZNF846_Imbeault2017_RP_ChIP-seq</td>\n",
       "      <td>0.06799531504110529</td>\n",
       "      <td>3.177141696287976</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('CDKL5', 0.5770311991890694), ('ELK3', 0.499...</td>\n",
       "      <td>1442</td>\n",
       "      <td>brca26-slice1</td>\n",
       "      <td>brca</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8384 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0         TF                                MotifID                   AUC  \\\n",
       "2       ATF4                      metacluster_148.1   0.04535987163002242   \n",
       "3       ATF7                      tfdimers__MD00403   0.07734985769355478   \n",
       "4       BRF1                     metacluster_172.30   0.07246964091972653   \n",
       "5      CEBPD                      metacluster_188.6   0.08619285687232511   \n",
       "6      CEBPD                      metacluster_127.2   0.08783092830360187   \n",
       "...      ...                                    ...                   ...   \n",
       "1247  ZNF704                      metacluster_144.3  0.052477081428336965   \n",
       "1248  ZNF707                       jaspar__MA1715.1   0.06133502145318984   \n",
       "1249  ZNF846                      metacluster_133.2   0.06818181818181818   \n",
       "1250  ZNF846      kznf__ZNF846_Imbeault2017_OM_MEME    0.0662720260209182   \n",
       "1251  ZNF846  kznf__ZNF846_Imbeault2017_RP_ChIP-seq   0.06799531504110529   \n",
       "\n",
       "0                    NES MotifSimilarityQvalue OrthologousIdentity  \\\n",
       "2      3.094187231280473                   0.0                 1.0   \n",
       "3     3.3327772737947217           2.25962e-06                 1.0   \n",
       "4      3.393020859843327                   0.0                 1.0   \n",
       "5     3.0759202055964154           4.11114e-06                 1.0   \n",
       "6      3.174015480466356           4.30447e-08                 1.0   \n",
       "...                  ...                   ...                 ...   \n",
       "1247   3.451496379493853                   0.0                 1.0   \n",
       "1248  3.2644445065952743                   0.0                 1.0   \n",
       "1249   3.193544830125133           2.55708e-06                 1.0   \n",
       "1250   3.025576739632648                   0.0                 1.0   \n",
       "1251   3.177141696287976                   0.0                 1.0   \n",
       "\n",
       "0                                            Annotation  \\\n",
       "2                            gene is directly annotated   \n",
       "3     motif similar to tfdimers__MD00119 ('M00973_re...   \n",
       "4                            gene is directly annotated   \n",
       "5     gene is annotated for similar motif tfdimers__...   \n",
       "6     gene is annotated for similar motif tfdimers__...   \n",
       "...                                                 ...   \n",
       "1247                         gene is directly annotated   \n",
       "1248                         gene is directly annotated   \n",
       "1249  motif similar to kznf__ZNF846_Imbeault2017_RP_...   \n",
       "1250                         gene is directly annotated   \n",
       "1251                         gene is directly annotated   \n",
       "\n",
       "0                                               Context  \\\n",
       "2     frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...   \n",
       "3     frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...   \n",
       "4     frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...   \n",
       "5     frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...   \n",
       "6     frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...   \n",
       "...                                                 ...   \n",
       "1247  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "1248  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "1249  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "1250  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "1251  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "\n",
       "0                                           TargetGenes RankAtMax  \\\n",
       "2     [('CYHR1', 0.7961399861048666), ('ATF3', 0.549...      4316   \n",
       "3     [('TBL1XR1', 0.3794181419999426), ('TP73', 0.9...      2860   \n",
       "4     [('APEH', 0.4085328157090424), ('KIAA0930', 0....      3481   \n",
       "5     [('PPM1H', 2.3367527585852383), ('WRAP73', 0.4...       920   \n",
       "6     [('WRAP73', 0.4242448518158827), ('HERPUD2', 0...       573   \n",
       "...                                                 ...       ...   \n",
       "1247  [('ZNF704', 1.0), ('SIAH1', 0.2095099932575139...      4645   \n",
       "1248  [('PIK3R1', 0.2066647993658202), ('ITSN2', 0.1...      3020   \n",
       "1249  [('CLSTN1', 0.3020024914065943), ('CCDC174', 0...       356   \n",
       "1250  [('SECTM1', 0.4894350004497896), ('CDKL5', 0.5...       381   \n",
       "1251  [('CDKL5', 0.5770311991890694), ('ELK3', 0.499...      1442   \n",
       "\n",
       "0             slice cancer  \n",
       "2     brca01-slice3   brca  \n",
       "3     brca01-slice3   brca  \n",
       "4     brca01-slice3   brca  \n",
       "5     brca01-slice3   brca  \n",
       "6     brca01-slice3   brca  \n",
       "...             ...    ...  \n",
       "1247  brca26-slice1   brca  \n",
       "1248  brca26-slice1   brca  \n",
       "1249  brca26-slice1   brca  \n",
       "1250  brca26-slice1   brca  \n",
       "1251  brca26-slice1   brca  \n",
       "\n",
       "[8384 rows x 12 columns]"
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancerlist=[\"brca\",\"cesc\",\"crc\",\"cscc\",\"gbm\",\"gist\",\"hn_as\",\"ipnm\",\"lihc\",\"luad\",\"mibc\",\"oscc\",\"ovca\",\"pcnsl\",\"pdac\",\"prad\",\"rcc\",\"skcm\"][0]\n",
    "dir_path=\"result_ctx\"\n",
    "select_slice_path=\"E:\\苏师兄代码和文件\\select_slice1.csv\"\n",
    "file_name=os.listdir(dir_path)\n",
    "file_path= [os.path.join(dir_path,i) for i in os.listdir(dir_path)]\n",
    "dfs = []\n",
    "select_slice=pd.read_csv(r\"E:\\苏师兄代码和文件\\select_slice1.csv\")[\"slice\"].values\n",
    "for i in file_name:\n",
    "    cancer=re.findall(r'\\D+', i.split(\"-\")[0])[0].replace(\"reg_\",\"\")\n",
    "    slice=i.split(\".loom.tsv.csv\")[0].replace(\"reg_\",\"\")\n",
    "    if cancer == \"brca\" and slice in select_slice:\n",
    "        data=pd.read_csv(r\"E:\\\\苏师兄代码和文件\\\\pyscenic\\\\result_ctx\\\\\"+i)\n",
    "        data.columns=data.iloc[0,:]\n",
    "        data=data[1:]\n",
    "        data.columns.values[0]=\"TF\"\n",
    "        data.columns.values[1]=\"MotifID\"\n",
    "        data=data[1:]\n",
    "        data[\"slice\"]=[slice for i in range(len(data.index))]\n",
    "        data[\"cancer\"]=[cancer for i in range(len(data.index))]\n",
    "        dfs.append(data)\n",
    "combined_df = pd.concat(dfs)\n",
    "table=pd.DataFrame(columns=[\"TF\",\"target\",\"target_num\",\"slice\",\"cancer\",\"NES\"])\n",
    "table[\"TF\"]=combined_df[\"TF\"]\n",
    "table[\"slice\"]=combined_df[\"slice\"]\n",
    "table[\"cancer\"]=combined_df[\"cancer\"]\n",
    "table[\"NES\"]=combined_df[\"NES\"]\n",
    "num=0\n",
    "Target_num=[]\n",
    "Target=[]\n",
    "for i,Str in enumerate(combined_df[\"TargetGenes\"]):\n",
    "    pattern = re.compile('[a-zA-Z]+')\n",
    "    target = [re.findall(pattern, item[0])[0] for item in eval(Str)]\n",
    "    Target_num.append(len(target))\n",
    "    Target.append(target)\n",
    "    table.iloc[i,1]=\",\".join(target)\n",
    "    num=num+1\n",
    "table[\"target_num\"]=Target_num\n",
    "table1=table.sort_values(by=\"target_num\",ascending=False)[:10]\n",
    "table1=table1.sort_values(by=\"NES\",ascending=False)[:10]\n",
    "table1.to_csv(\"Top10_sort_by_NES.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancerlist=[\"brca\",\"cesc\",\"crc\",\"cscc\",\"gbm\",\"gist\",\"hn_as\",\"ipnm\",\"lihc\",\"luad\",\"mibc\",\"oscc\",\"ovca\",\"pcnsl\",\"pdac\",\"prad\",\"rcc\",\"skcm\"][0]\n",
    "dir_path=\"result_ctx\"\n",
    "select_slice_path=\"E:\\苏师兄代码和文件\\select_slice1.csv\"\n",
    "file_name=os.listdir(dir_path)\n",
    "file_path= [os.path.join(dir_path,i) for i in os.listdir(dir_path)]\n",
    "dfs = []\n",
    "select_slice=pd.read_csv(r\"E:\\苏师兄代码和文件\\select_slice1.csv\")[\"slice\"].values\n",
    "for i in file_name:\n",
    "    cancer=re.findall(r'\\D+', i.split(\"-\")[0])[0].replace(\"reg_\",\"\")\n",
    "    slice=i.split(\".loom.tsv.csv\")[0].replace(\"reg_\",\"\")\n",
    "    if cancer == \"brca\" and slice in select_slice:\n",
    "        data=pd.read_csv(r\"E:\\\\苏师兄代码和文件\\\\pyscenic\\\\result_ctx\\\\\"+i)\n",
    "        data.columns=data.iloc[0,:]\n",
    "        data=data[1:]\n",
    "        data.columns.values[0]=\"TF\"\n",
    "        data.columns.values[1]=\"MotifID\"\n",
    "        data=data[1:]\n",
    "        data[\"slice\"]=[slice for i in range(len(data.index))]\n",
    "        data[\"cancer\"]=[cancer for i in range(len(data.index))]\n",
    "        dfs.append(data)\n",
    "combined_df = pd.concat(dfs)\n",
    "table=pd.DataFrame(columns=[\"TF\",\"target\",\"target_num\",\"slice\",\"cancer\",\"AUC\"])\n",
    "table[\"TF\"]=combined_df[\"TF\"]\n",
    "table[\"slice\"]=combined_df[\"slice\"]\n",
    "table[\"cancer\"]=combined_df[\"cancer\"]\n",
    "table[\"AUC\"]=combined_df[\"AUC\"]\n",
    "num=0\n",
    "Target_num=[]\n",
    "Target=[]\n",
    "for i,Str in enumerate(combined_df[\"TargetGenes\"]):\n",
    "    pattern = re.compile('[a-zA-Z]+')\n",
    "    target = [re.findall(pattern, item[0])[0] for item in eval(Str)]\n",
    "    Target_num.append(len(target))\n",
    "    Target.append(target)\n",
    "    table.iloc[i,1]=\",\".join(target)\n",
    "    num=num+1\n",
    "table[\"target_num\"]=Target_num\n",
    "table1=table.sort_values(by=\"target_num\",ascending=False)[:10]\n",
    "table1=table1.sort_values(by=\"AUC\",ascending=False)[:10]\n",
    "table1.to_csv(\"Top10_sort_by_AUC.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>MotifID</th>\n",
       "      <th>AUC</th>\n",
       "      <th>NES</th>\n",
       "      <th>MotifSimilarityQvalue</th>\n",
       "      <th>OrthologousIdentity</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Context</th>\n",
       "      <th>TargetGenes</th>\n",
       "      <th>RankAtMax</th>\n",
       "      <th>slice</th>\n",
       "      <th>cancer</th>\n",
       "      <th>Target_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATF4</td>\n",
       "      <td>metacluster_148.1</td>\n",
       "      <td>0.04535987163002242</td>\n",
       "      <td>3.094187231280473</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...</td>\n",
       "      <td>[('CYHR1', 0.7961399861048666), ('ATF3', 0.549...</td>\n",
       "      <td>4316</td>\n",
       "      <td>brca01-slice3</td>\n",
       "      <td>b</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ATF7</td>\n",
       "      <td>tfdimers__MD00403</td>\n",
       "      <td>0.07734985769355478</td>\n",
       "      <td>3.3327772737947217</td>\n",
       "      <td>2.25962e-06</td>\n",
       "      <td>1.0</td>\n",
       "      <td>motif similar to tfdimers__MD00119 ('M00973_re...</td>\n",
       "      <td>frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...</td>\n",
       "      <td>[('TBL1XR1', 0.3794181419999426), ('TP73', 0.9...</td>\n",
       "      <td>2860</td>\n",
       "      <td>brca01-slice3</td>\n",
       "      <td>b</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRF1</td>\n",
       "      <td>metacluster_172.30</td>\n",
       "      <td>0.07246964091972653</td>\n",
       "      <td>3.393020859843327</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...</td>\n",
       "      <td>[('APEH', 0.4085328157090424), ('KIAA0930', 0....</td>\n",
       "      <td>3481</td>\n",
       "      <td>brca01-slice3</td>\n",
       "      <td>b</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CEBPD</td>\n",
       "      <td>metacluster_188.6</td>\n",
       "      <td>0.08619285687232511</td>\n",
       "      <td>3.0759202055964154</td>\n",
       "      <td>4.11114e-06</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is annotated for similar motif tfdimers__...</td>\n",
       "      <td>frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...</td>\n",
       "      <td>[('PPM1H', 2.3367527585852383), ('WRAP73', 0.4...</td>\n",
       "      <td>920</td>\n",
       "      <td>brca01-slice3</td>\n",
       "      <td>b</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CEBPD</td>\n",
       "      <td>metacluster_127.2</td>\n",
       "      <td>0.08783092830360187</td>\n",
       "      <td>3.174015480466356</td>\n",
       "      <td>4.30447e-08</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is annotated for similar motif tfdimers__...</td>\n",
       "      <td>frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...</td>\n",
       "      <td>[('WRAP73', 0.4242448518158827), ('HERPUD2', 0...</td>\n",
       "      <td>573</td>\n",
       "      <td>brca01-slice3</td>\n",
       "      <td>b</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1247</th>\n",
       "      <td>ZNF704</td>\n",
       "      <td>metacluster_144.3</td>\n",
       "      <td>0.052477081428336965</td>\n",
       "      <td>3.451496379493853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('ZNF704', 1.0), ('SIAH1', 0.2095099932575139...</td>\n",
       "      <td>4645</td>\n",
       "      <td>brca26-slice1</td>\n",
       "      <td>b</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1248</th>\n",
       "      <td>ZNF707</td>\n",
       "      <td>jaspar__MA1715.1</td>\n",
       "      <td>0.06133502145318984</td>\n",
       "      <td>3.2644445065952743</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('PIK3R1', 0.2066647993658202), ('ITSN2', 0.1...</td>\n",
       "      <td>3020</td>\n",
       "      <td>brca26-slice1</td>\n",
       "      <td>b</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1249</th>\n",
       "      <td>ZNF846</td>\n",
       "      <td>metacluster_133.2</td>\n",
       "      <td>0.06818181818181818</td>\n",
       "      <td>3.193544830125133</td>\n",
       "      <td>2.55708e-06</td>\n",
       "      <td>1.0</td>\n",
       "      <td>motif similar to kznf__ZNF846_Imbeault2017_RP_...</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('CLSTN1', 0.3020024914065943), ('CCDC174', 0...</td>\n",
       "      <td>356</td>\n",
       "      <td>brca26-slice1</td>\n",
       "      <td>b</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1250</th>\n",
       "      <td>ZNF846</td>\n",
       "      <td>kznf__ZNF846_Imbeault2017_OM_MEME</td>\n",
       "      <td>0.0662720260209182</td>\n",
       "      <td>3.025576739632648</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('SECTM1', 0.4894350004497896), ('CDKL5', 0.5...</td>\n",
       "      <td>381</td>\n",
       "      <td>brca26-slice1</td>\n",
       "      <td>b</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1251</th>\n",
       "      <td>ZNF846</td>\n",
       "      <td>kznf__ZNF846_Imbeault2017_RP_ChIP-seq</td>\n",
       "      <td>0.06799531504110529</td>\n",
       "      <td>3.177141696287976</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>gene is directly annotated</td>\n",
       "      <td>frozenset({'activating', 'top50perTarget', 'hg...</td>\n",
       "      <td>[('CDKL5', 0.5770311991890694), ('ELK3', 0.499...</td>\n",
       "      <td>1442</td>\n",
       "      <td>brca26-slice1</td>\n",
       "      <td>b</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8384 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0         TF                                MotifID                   AUC  \\\n",
       "2       ATF4                      metacluster_148.1   0.04535987163002242   \n",
       "3       ATF7                      tfdimers__MD00403   0.07734985769355478   \n",
       "4       BRF1                     metacluster_172.30   0.07246964091972653   \n",
       "5      CEBPD                      metacluster_188.6   0.08619285687232511   \n",
       "6      CEBPD                      metacluster_127.2   0.08783092830360187   \n",
       "...      ...                                    ...                   ...   \n",
       "1247  ZNF704                      metacluster_144.3  0.052477081428336965   \n",
       "1248  ZNF707                       jaspar__MA1715.1   0.06133502145318984   \n",
       "1249  ZNF846                      metacluster_133.2   0.06818181818181818   \n",
       "1250  ZNF846      kznf__ZNF846_Imbeault2017_OM_MEME    0.0662720260209182   \n",
       "1251  ZNF846  kznf__ZNF846_Imbeault2017_RP_ChIP-seq   0.06799531504110529   \n",
       "\n",
       "0                    NES MotifSimilarityQvalue OrthologousIdentity  \\\n",
       "2      3.094187231280473                   0.0                 1.0   \n",
       "3     3.3327772737947217           2.25962e-06                 1.0   \n",
       "4      3.393020859843327                   0.0                 1.0   \n",
       "5     3.0759202055964154           4.11114e-06                 1.0   \n",
       "6      3.174015480466356           4.30447e-08                 1.0   \n",
       "...                  ...                   ...                 ...   \n",
       "1247   3.451496379493853                   0.0                 1.0   \n",
       "1248  3.2644445065952743                   0.0                 1.0   \n",
       "1249   3.193544830125133           2.55708e-06                 1.0   \n",
       "1250   3.025576739632648                   0.0                 1.0   \n",
       "1251   3.177141696287976                   0.0                 1.0   \n",
       "\n",
       "0                                            Annotation  \\\n",
       "2                            gene is directly annotated   \n",
       "3     motif similar to tfdimers__MD00119 ('M00973_re...   \n",
       "4                            gene is directly annotated   \n",
       "5     gene is annotated for similar motif tfdimers__...   \n",
       "6     gene is annotated for similar motif tfdimers__...   \n",
       "...                                                 ...   \n",
       "1247                         gene is directly annotated   \n",
       "1248                         gene is directly annotated   \n",
       "1249  motif similar to kznf__ZNF846_Imbeault2017_RP_...   \n",
       "1250                         gene is directly annotated   \n",
       "1251                         gene is directly annotated   \n",
       "\n",
       "0                                               Context  \\\n",
       "2     frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...   \n",
       "3     frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...   \n",
       "4     frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...   \n",
       "5     frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...   \n",
       "6     frozenset({'hg38_10kbp_up_10kbp_down_full_tx_v...   \n",
       "...                                                 ...   \n",
       "1247  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "1248  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "1249  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "1250  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "1251  frozenset({'activating', 'top50perTarget', 'hg...   \n",
       "\n",
       "0                                           TargetGenes RankAtMax  \\\n",
       "2     [('CYHR1', 0.7961399861048666), ('ATF3', 0.549...      4316   \n",
       "3     [('TBL1XR1', 0.3794181419999426), ('TP73', 0.9...      2860   \n",
       "4     [('APEH', 0.4085328157090424), ('KIAA0930', 0....      3481   \n",
       "5     [('PPM1H', 2.3367527585852383), ('WRAP73', 0.4...       920   \n",
       "6     [('WRAP73', 0.4242448518158827), ('HERPUD2', 0...       573   \n",
       "...                                                 ...       ...   \n",
       "1247  [('ZNF704', 1.0), ('SIAH1', 0.2095099932575139...      4645   \n",
       "1248  [('PIK3R1', 0.2066647993658202), ('ITSN2', 0.1...      3020   \n",
       "1249  [('CLSTN1', 0.3020024914065943), ('CCDC174', 0...       356   \n",
       "1250  [('SECTM1', 0.4894350004497896), ('CDKL5', 0.5...       381   \n",
       "1251  [('CDKL5', 0.5770311991890694), ('ELK3', 0.499...      1442   \n",
       "\n",
       "0             slice cancer  Target_num  \n",
       "2     brca01-slice3      b         106  \n",
       "3     brca01-slice3      b          31  \n",
       "4     brca01-slice3      b          32  \n",
       "5     brca01-slice3      b          11  \n",
       "6     brca01-slice3      b           8  \n",
       "...             ...    ...         ...  \n",
       "1247  brca26-slice1      b         186  \n",
       "1248  brca26-slice1      b          32  \n",
       "1249  brca26-slice1      b          13  \n",
       "1250  brca26-slice1      b          12  \n",
       "1251  brca26-slice1      b          27  \n",
       "\n",
       "[8384 rows x 13 columns]"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combined_df[\"TargetGenes_name\"]=english_words\n",
    "num=0\n",
    "\n",
    "\n",
    "for group_name, group_data in group_list:\n",
    "    target=[i for i in group_data[\"target\"]]\n",
    "    importance=[str(i) for i in group_data[\"importance\"].round(3)]\n",
    "\n",
    "    cancer=[i for i in group_data[\"cancer\"]]\n",
    "    target_num=len(target)\n",
    "    table.loc[num,\"TF\"]=group_name[0]\n",
    "    table.loc[num,\"target\"]=\",\".join(target)\n",
    "    table.loc[num,\"target_num\"]=target_num\n",
    "    table.loc[num,\"importance\"]=\",\".join(importance)\n",
    "    table.loc[num,\"slice\"]=group_name[1].replace(\"-\",\"_\")\n",
    "    table.loc[num,\"cancer\"]=self.cancer\n",
    "    num+=1\n",
    "table.to_csv(self.save_path+\"\\\\TF-Statistics\"+\".txt\",sep=\"\\t\",index=False)\n",
    "print(f\"{self.cancer}的TF-target统计完成了!\")\n",
    "print(\"******************************************************************************************************************\")\n",
    "print(table)\n",
    "        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
